x-pack/plugins/ml/server/models/data_recognizer/data_recognizer.js

```diff
@@ -583,6 +588,16 @@ export class DataRecognizer {
             jsonString = jsonString.replace(new RegExp(INDEX_PATTERN_ID, 'g'), this.indexPatternId);
             item.config.kibanaSavedObjectMeta.searchSourceJSON = jsonString;
           }
+
+          if (category === SAVED_OBJECT_TYPES.VISUALIZATION) {
+            // Look for any INDEX_PATTERN_NAME tokens in visualization visState,
+            // as e.g. Vega visualizations reference the Elasticsearch index pattern directly.
+            let visStateString = item.config.visState;
+            if (visStateString !== undefined && visStateString.match(INDEX_PATTERN_NAME)) {
+              visStateString = visStateString.replace(new RegExp(INDEX_PATTERN_NAME, 'g'), this.indexPatternName);
```

Comment:
there might be a risk here that we'll match more than just the index pattern name.
as the index pattern name is defined by the user, it could be anything, and could be a word that coincidentally matches in the `visStateString`. e.g, if the index pattern was `random_score`, this would corrupt the saved object, as that word also exists in the json.
we could qualify the match with a partial string that we know will always be included along side it e.g.  `index:${INDEX_PATTERN_NAME}"`